vhost-device-sound/src/device.rs

```diff
@@ -495,77 +437,35 @@ impl VhostUserSoundThread {
-                        if descriptor.len() as usize == size_of::<VirtioSoundPcmStatus>() =>
-                    {
-                        self.streams.write().unwrap()[stream_id as usize]
-                            .buffers
-                            .extend(std::mem::take(&mut buffers).into_iter());
-                        state = IoState::Done;
-                    }
-                    IoState::Ready
-                        if descriptor.len() as usize != size_of::<VirtioSoundPcmXfer>() =>
-                    {
-                        return Err(Error::UnexpectedDescriptorSize(
-                            size_of::<VirtioSoundPcmXfer>(),
-                            descriptor.len(),
-                        )
-                        .into());
-                    }
-                    IoState::Ready => {
-                        let xfer = desc_chain
-                            .memory()
-                            .read_obj::<VirtioSoundPcmXfer>(descriptor.addr())
-                            .map_err(|_| Error::DescriptorReadFailed)?;
-                        let stream_id: u32 = xfer.stream_id.into();
-                        stream_ids.insert(stream_id);
-
-                        state = IoState::WaitingBufferForStreamId(stream_id);
-                    }
-                    IoState::WaitingBufferForStreamId(stream_id)
-                        if descriptor.len() as usize == size_of::<VirtioSoundPcmXfer>() =>
-                    {
-                        return Err(Error::UnexpectedDescriptorSize(
-                            u32::from(
-                                self.streams.read().unwrap()[stream_id as usize]
-                                    .params
-                                    .period_bytes,
-                            ) as usize,
-                            descriptor.len(),
-                        )
-                        .into());
-                    }
-                    IoState::WaitingBufferForStreamId(_) => {
-                        // In the case of TX/Playback:
-                        //
-                        // Rather than copying the content of a descriptor, buffer keeps a pointer
-                        // to it. When we copy just after the request is enqueued, the guest's
-                        // userspace may or may not have updated the buffer contents. Guest driver
-                        // simply moves buffers from the used ring to the available ring without
-                        // knowing whether the content has been updated. The device only reads the
-                        // buffer from guest memory when the audio engine requires it, which is
-                        // about after a period thus ensuring that the buffer is up-to-date.
-                        buffers.push(Buffer::new(*descriptor, Arc::clone(&message), direction));
-                    }
+
+            let mem = atomic_mem.memory();
```

Comment:
maybe it's better to keep `atomic_mem.memory()` outside the loop to avoid repetitive calls.